sql/catalyst/src/main/scala/org/apache/spark/sql/types/Metadata.scala

```diff
@@ -104,7 +104,8 @@ sealed class Metadata private[types] (private[types] val map: Map[String, Any])
     }
   }
 
-  override def hashCode: Int = Metadata.hash(this)
+  private lazy val _hashCode: Int = Metadata.hash(this)
+  override def hashCode: Int = _hashCode
```

Comment:
Making `hashCode` into a `lazy val` triggers compilation errors and triggers Scalastyle's "You should implement equals and hashCode together" error; it also causes problems in Scala code which calls `.hashCode()`:

```
[error] /Users/joshrosen/Documents/spark/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/namedExpressions.scala:247: type mismatch;
[error]  found   : AttributeReference.this.metadata.type (with underlying type org.apache.spark.sql.types.Metadata)
[error]  required: ?{def hashCode(): ?}
[error] Note that Metadata extends Any, not AnyRef.
[error] Such types can participate in value classes, but instances
[error] cannot appear in singleton types or in reference comparisons.
[error]     h = h * 37 + metadata.hashCode()
[error]                  ^
[error] /Users/joshrosen/Documents/spark/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/namedExpressions.scala:247: Int does not take parameters
[error]     h = h * 37 + metadata.hashCode()
[error]                                   ^
[error] two errors found
[error] /Users/joshrosen/Documents/spark/sql/catalyst/src/main/scala/org/apache/spark/sql/types/Metadata.scala:41:13: You should implement equals and hashCode together
```